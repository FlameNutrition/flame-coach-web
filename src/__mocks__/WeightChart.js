// FIXME: Try to understand why __mock__ folder doesn't work
const WeightChart = () => 'test';

export default WeightChart;
