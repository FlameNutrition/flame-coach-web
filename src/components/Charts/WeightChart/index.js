import PropTypes from 'prop-types';
import React, { useState } from 'react';
import clsx from 'clsx';
import '../../../../node_modules/react-vis/dist/style.css';
import {
  FlexibleXYPlot,
  Hint,
  HorizontalGridLines,
  LineMarkSeries,
  VerticalGridLines,
  XAxis,
  YAxis
} from 'react-vis';
import moment from 'moment';
import { maxTicksLimit, maxTicksLimitMobile, minMaxValue } from '../utils/chartUtil';
import { getTimeRangeDates } from '../../../utils/timeRangeUtil';
import Card from '@material-ui/core/Card';
import makeStyles from '@material-ui/styles/makeStyles';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';

const useStyles = makeStyles(() => ({
  root: {},
  chart: {
    display: 'flex',
    height: '100%',
    width: '100%'
  },
  weightGraphicTitle: {
    textAlign: 'center',
    paddingBottom: '10px'
  }
}));

const WeightChart = ({
  isMobile,
  timeFrame,
  className,
  dataChart,
  measureUnit
}) => {
  const classes = useStyles();

  const maxTick = isMobile ? maxTicksLimitMobile(timeFrame) : maxTicksLimit(timeFrame);

  const filterDates = getTimeRangeDates(timeFrame, moment());

  const minMaxWeightRange = minMaxValue(dataChart);

  const dataFormatted = dataChart.map((data) => ({
    y: data.value,
    x: moment(data.date, 'YYYY-MM-DD')
  }));

  const fromDttm = moment(filterDates.fromDttm, 'YYYY-MM-DD');
  const toDttm = moment(filterDates.toDttm, 'YYYY-MM-DD');

  const minWeight = minMaxWeightRange.minValue - 10;
  const maxWeight = minMaxWeightRange.maxValue + 10;

  const [value, setValue] = useState();

  return (
    <Card className={clsx(classes.root)}>
      <CardContent className={className}>
        <Typography className={classes.weightGraphicTitle} component="h2" variant="h5" gutterBottom>
          Weight
        </Typography>
        <Box className={classes.chart}>
          <FlexibleXYPlot
            margin={{
              left: 30,
              right: 30,
              top: 5,
              bottom: isMobile ? 30 : 60
            }}
            xType="time"
            yDomain={[minWeight, maxWeight]}
            xDomain={[fromDttm, toDttm]}
          >
            <VerticalGridLines/>
            <HorizontalGridLines/>
            <XAxis
              tickTotal={maxTick}
              tickSize={0}
              tickPadding={isMobile ? 12 : 10}
              tickFormat={(xValue) => {
                return moment(xValue)
                  .format('MM-DD');
              }}
              tickLabelAngle={isMobile ? -45 : 0}
              style={{
                ticks: { paddingLeft: '50px' },
                text: { fontFamily: '"Roboto", "Helvetica", "Arial", sans-serif' }
              }}
            />
            <YAxis
              tickSizeOuter={0}
              tickSizeInner={0}
              tickPadding={7}
              style={{
                text: { fontFamily: '"Roboto", "Helvetica", "Arial", sans-serif' }
              }}
            />
            <LineMarkSeries
              data={dataFormatted}
              size={3}
              color="#3f51b5"
              fill="#3f51b5"
              onValueMouseOver={(datapoint) => {
                setValue(datapoint);
              }}
              onValueMouseOut={() => setValue(null)}
            />
            {value ? (
              <Hint
                value={value}
                format={(datapoint) => {
                  return [
                    {
                      title: 'Date',
                      value: moment(datapoint.x)
                        .format('YYYY-MM-DD')
                    },
                    {
                      title: 'Weight',
                      value: `${datapoint.y} ${measureUnit}`
                    }];
                }}
                style={{ text: { fontFamily: '"Roboto", "Helvetica", "Arial", sans-serif' } }}
              />
            ) : null}
          </FlexibleXYPlot>
        </Box>
      </CardContent>
    </Card>
  );
};

WeightChart.propTypes = {
  className: PropTypes.string,
  timeFrame: PropTypes.string.isRequired,
  dataChart: PropTypes.array.isRequired,
  isMobile: PropTypes.bool,
  measureUnit: PropTypes.string.isRequired
};

WeightChart.defaultProps = {};

export default WeightChart;
