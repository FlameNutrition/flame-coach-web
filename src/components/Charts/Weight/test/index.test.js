import React from 'react';
import WeightChart from '../index';
import { render } from '../../../../testing/test-utils';

// Is not possible to test the canvas
jest.mock('react-chartjs-2', () => ({
  Line: () => null,
}));

describe('<WeightChart />', () => {
  it('create component WeightChart', () => {
    const { container } = render(<WeightChart
      dataChart={[60, 70, 50, 70, 50, 30, 80]}
      timeFrame="1_WEEK"
    />);

    expect(container).toMatchSnapshot();
  });
});
