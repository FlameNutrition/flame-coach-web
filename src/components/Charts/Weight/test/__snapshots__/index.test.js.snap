// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WeightChart /> create component WeightChart 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root"
    >
      <h2
        class="MuiTypography-root makeStyles-weightGraphicTitle-1 MuiTypography-h5 MuiTypography-gutterBottom"
      >
        Weight
      </h2>
    </div>
  </div>
</div>
`;
